craftdb.lua

```diff
@@ -214,28 +226,84 @@ function CraftDB:get_all_recipes(item_list)
 end
 
 
--- TODO: Add filtering so we can exclude nodes made via tablesaw.
-function CraftDB:find_all_matching_items(name_pattern, offset, max_count)
+function CraftDB:search_items(name_pattern, options)
   local matching_names = {}  -- list of names
-  local image_names = {}  -- name -> image_filename
 
   if type(name_pattern) ~= 'string' then name_pattern = '.' end
 
-  if type(offset) ~= 'number' then offset = 1 end
-  offset = math.max(1, offset)
+  if type(options) ~= 'table' then options = {} end
+
+  local offset = (type(options.offset) == 'number') and
+                  math.max(1, options.offset) or 1
+
+  local max_count = (type(options.max_count) == 'number') and
+                  math.min(math.max(1, options.max_count), MAX_MATCHES) or
+                  MAX_MATCHES
+
+  -- Generate exclusion lookup tables from the input lists.
+  local exclude_groups = _make_lut_from_list(options['exclude_groups'])
```

Comment:
Allowing directly supplied lookup table would allow more complex filters.
Not directly requesting change here as this probably belongs to corner cases category but could consider it to remove need of reconstructing exclude_groups table and to allow complex filters like searching tools based on weapon damage or hardness of node.
Doable later too if needed but bit harder after API is stable as it would require more code to keep backward compatibility with indexed tables.

Reply:
Also if going for this it would still need ability to do truth value check to allow simple filtering, that would probably be `group_name = true` to accept any group value that is present and not 0.

Reply:
Do you have a suggested syntax of what the actual usage would be?  How would a player specify "I want all normal wood planks that are in group:wood" vs "I want to see all blocks that are not oddly_breakable_by_hand and can only be broken with a diamond pickaxe"?

I suppose at some point we should just return the table to the user and let them filter it client-side (eg, in the LUAC), but there are practical limits to how big a message LUAC can cache in mem and process before timing out.

Reply:
> Do you have a suggested syntax of what the actual usage would be?

Yes:
 ```lua
{
   filter_groups_or_whatever_you_call_it = {
     oddly_breakable_by_hand = 1,
     cracky = 3,
  },
}
```
Which should match things that are in `oddly_breakable_by_hand` group and has rating 3 in `cracky` group.

However, like stated before this would also have to allow fuzzy match with just truth value:
```lua
{
   filter_groups_or_whatever_you_call_it = {
     oddly_breakable_by_hand = true,
     cracky = true,
  },
}
```
Which should match things that are in `oddly_breakable_by_hand` group and has any rating in `cracky` group.

I mean there's no reason to implement your own specification and convert it in API to compatible spec when you could achieve more by allowing user to directly supply groups like how engine actually uses them.

Also response contains group in that same format, this allows repeating searches using response data directly.
Also in my opinion it is bad to have different format in query and response for same thing especially if it does not bring any significant improvements to API.
Additional less important value is that syntax is easier to remember for modders and also can help non modders to learn how minetest groups work.

Reply:
I have updated the PR to address this feedback, but not the others yet.  Please review this section if you have time.  TYVM.